src/slice-chop.c

```diff
@@ -612,9 +612,28 @@ static SEXP vec_as_indices(SEXP indices, R_len_t n, SEXP names) {
 
   R_len_t size = vec_size(indices);
 
+  // Restrict index values to positive integer locations
+  struct vec_as_location_opts opts = {
+    .action = SUBSCRIPT_ACTION_DEFAULT,
+    .missing = SUBSCRIPT_MISSING_PROPAGATE,
+    .loc_negative = LOC_NEGATIVE_ERROR,
+    .loc_oob = LOC_OOB_ERROR,
+    .loc_zero = LOC_ZERO_ERROR,
+    .subscript_arg = R_NilValue
+  };
+
   for (int i = 0; i < size; ++i) {
     index = VECTOR_ELT(indices, i);
-    SET_VECTOR_ELT(indices, i, vec_as_location(index, n, names));
+
+    // Restrict the type to only numeric values.
```

Comment:
This is not completely right, for instance factors are implemented as numeric. We should check for OBJECT and call `vec_as_subscript()` for S3 inputs.

Reply:
I need to fix a similar problem for #575. I'll look into a C-level `vec_as_subscript()`.

Reply:
I'm not sure I want S3 inputs to be allowed here. I really think I just want `num_as_location()`, but I'm now seeing that even that allows S3's

``` r
vctrs::num_as_location(factor("x"), 1)
#> [1] x
#> Levels: x
```

<sup>Created on 2020-02-25 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

Reply:
Not accepting S3 inputs would be inconsistent with other subscript handling. If we take locations we should take any type coercible to integer.

Reply:
I'll fix the num-as-location bug.